src/architecture/architect.js

```diff
@@ -5,562 +5,596 @@ var Layer = require('./layer');
-      let j;
-      if (list[i] instanceof Group) {
-        for (j = 0; j < list[i].nodes.length; j++) {
-          nodes.push(list[i].nodes[j]);
-        }
-      } else if (list[i] instanceof Layer) {
-        for (j = 0; j < list[i].nodes.length; j++) {
-          for (var k = 0; k < list[i].nodes[j].nodes.length; k++) {
-            nodes.push(list[i].nodes[j].nodes[k]);
-          }
+/**
+* Constructs a network from a given array of connected nodes
+*
+* @param {Group[]|Layer[]|Node[]} list A list of Groups, Layers, and Nodes to combine into a Network
+*
+* @example <caption>A Network built with Nodes</caption>
+* let { architect } = require("@liquid-carrot/carrot");
+*
+* var A = new Node();
+* var B = new Node();
+* var C = new Node();
+* var D = new Node();
+*
+* // Create connections
+* A.connect(B);
+* A.connect(C);
+* B.connect(D);
+* C.connect(D);
+*
+* // Construct a network
+* var network = architect.Construct([A, B, C, D]);
+*
+* @example <caption>A Network built with Groups</caption>
+* let { architect } = require("@liquid-carrot/carrot");
+*
+* var A = new Group(4);
+* var B = new Group(2);
+* var C = new Group(6);
+*
+* // Create connections between the groups
+* A.connect(B);
+* A.connect(C);
+* B.connect(C);
+*
+* // Construct a square-looking network
+* var network = architect.Construct([A, B, C, D]);
+*
+* @returns {Network}
+*/
+Construct: function (list) {
+  // Create a network
+  var network = new Network(0, 0);
+
+  // Transform all groups into nodes
+  var nodes = [];
+
+  var i;
+  for (i = 0; i < list.length; i++) {
+    let j;
+    if (list[i] instanceof Group) {
+      for (j = 0; j < list[i].nodes.length; j++) {
+        nodes.push(list[i].nodes[j]);
+      }
+    } else if (list[i] instanceof Layer) {
+      for (j = 0; j < list[i].nodes.length; j++) {
+        for (var k = 0; k < list[i].nodes[j].nodes.length; k++) {
+          nodes.push(list[i].nodes[j].nodes[k]);
         }
-      } else if (list[i] instanceof Node) {
-        nodes.push(list[i]);
       }
+    } else if (list[i] instanceof Node) {
+      nodes.push(list[i]);
     }
+  }
 
-    // Determine input and output nodes
-    var inputs = [];
-    var outputs = [];
-    for (i = nodes.length - 1; i >= 0; i--) {
-      if (nodes[i].type === 'output' || nodes[i].connections.out.length + nodes[i].connections.gated.length === 0) {
-        nodes[i].type = 'output';
-        network.output++;
-        outputs.push(nodes[i]);
-        nodes.splice(i, 1);
-      } else if (nodes[i].type === 'input' || !nodes[i].connections.in.length) {
-        nodes[i].type = 'input';
-        network.input++;
-        inputs.push(nodes[i]);
-        nodes.splice(i, 1);
-      }
-    }
+  // Determine input and output nodes
+  var inputs = [];
+  var outputs = [];
+  for (i = nodes.length - 1; i >= 0; i--) {
+
+   if (nodes[i].type === 'output' || nodes[i].connections.in.length && nodes[i].connections.out.length + nodes[i].connections.gated.length === 0) {
```

Comment:
A new null check is added here. But is it `nodes[i].connections.in.length` that is nullable or is it `nodes[i].connections.in`? Does the same check need to happen for `nodes[i].connections.out`?

Reply:
Hey Dan! Good catch! There probably should be an equivalent test for the input case where `else if (nodes[i].type === 'input' || nodes[i].connections.out.length && !nodes[i].connections.in.length)`